### PR TITLE
refactor: use declarative state for executing mutations

### PR DESCRIPTION
**Pull request by sborrazas**
The goal of these changes is to be able to use a interchangeable state
that can perform these executions. The functional way of dealing with
any sort of state mutations allows this, and which will also allow
performing in-memory updates to the state.

### CODE REVIEW

File: lib/ae_mdw/db/sync/block.ex
Comment (jyeshe) [inline lib/ae_mdw/db/sync/block.ex:71]: `from_mbi != 0` maybe? (for semantics)

File: lib/ae_mdw/sync/server.ex
Comment (jyeshe) [inline lib/ae_mdw/sync/server.ex:167]: After key block commit it would be a time also for `Broadcaster.broadcast_key_block(key_block, :mdw)`

File: lib/ae_mdw/sync/server.ex
Comment (jyeshe) [inline lib/ae_mdw/sync/server.ex:134]: A nitpick as there is `from_txi` and `to_height` maybe instead of `block_index` could we separate it as `from_height` and `from_mbi`?

File: lib/ae_mdw/sync/server.ex
Comment (jyeshe) [inline lib/ae_mdw/sync/server.ex:139]: There are some blocks that take a long time to load them from node DB and dry-run the events. What you think of having commits for every height?

**Review by jyeshe — COMMENTED** [review]
Very neat PR! :100: Only one topic to analyse an alternative. If it would be possible to have commits for every height.

File: lib/ae_mdw/db/origin.ex
Comment (sborrazas) [reply lib/ae_mdw/db/origin.ex:64]: Good idea, done, added it to the cache

File: lib/ae_mdw/db/sync/block.ex
Comment (sborrazas) [reply lib/ae_mdw/db/sync/block.ex:71]: If we don't check for `from_height  == height` then we would be skipping `from_mbi` blocks for every height

File: lib/ae_mdw/sync/server.ex
Comment (sborrazas) [reply lib/ae_mdw/sync/server.ex]: Added it, thanks!

File: lib/ae_mdw/sync/server.ex
Comment (sborrazas) [reply lib/ae_mdw/sync/server.ex]: Done

File: lib/ae_mdw/sync/server.ex
Comment (sborrazas) [reply lib/ae_mdw/sync/server.ex:139]: That's a good idea, I'd like to add that on a following PR so we can test it thoroughly from scratch

File: lib/ae_mdw/db/sync/block.ex
Comment (jyeshe) [reply lib/ae_mdw/db/sync/block.ex:71]: At least for readability, looks like something can be improved, maybe putting both conditions:
`from_height == height and from_mbi != 0` but even after it the `from_height == height` part (from my perspective of not writing the code) would continue to be not easily readable.

Comment (sborrazas) [reply lib/ae_mdw/db/sync/block.ex]: Done

**Comment by sborrazas** [comment]
These final version is now passing the integration tests + diff script. There some differences in the stats which I think is now the correct version. Ready for review new

**Review by thepiwo — APPROVED** [review]

**Comment by thepiwo** [comment]
whats the state of this, I think we discussed not to have PRs hanging for so long generally

**Review by jyeshe — COMMENTED** [review]
After unit tests passing, it will be approved provided that as discussed there will be a following PR that makes commits for each height.

**Review by jyeshe — APPROVED** [review]

**Pull request merged**
